In [1]:
.libPaths(R.home('/opt/homebrew/Caskroom/miniforge/base/lib/R/library'))

In [2]:
Output = '/Users/alexis/Library/CloudStorage/OneDrive-UniversityofNorthCarolinaatChapelHill/CEMALB_DataAnalysisPM/Projects/P1015. Fire Sufficient Similarity/P1015.3. Analyses/P1015.3.2. Data Processing/Output'
cur_date = '061225'

library(missForest)
library(readxl)
library(openxlsx)
library(tidyverse)
library(factoextra)

# reading in files
ws_df = data.frame(Sample = 'WS', read_excel("Input/Woodsmoke_Data_053125.xlsx", sheet = 2)) %>%
    select(-Number)
wf_df = data.frame(Sample = 'WF', read_excel("Input/Wildfire_Data_053125.xlsx", sheet = 2)) %>%
    select(-Number)

── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.5
✔ forcats   1.0.0     ✔ stringr   1.5.1
✔ ggplot2   3.5.2     ✔ tibble    3.2.1
✔ lubridate 1.9.4     ✔ tidyr     1.3.1
✔ purrr     1.0.4     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors
Welcome! Want to learn more? See two factoextra-related books at https://goo.gl/ve3WBa



In [3]:
head(ws_df)
head(wf_df)

Sample HAWC_ID Study            Sample_Number Class Metric DTXSID        
1 WS     429174  Arif et al. 2017 1             PAH   Weight DTXSID6025648 
2 WS     429174  Arif et al. 2017 1             PAH   Weight DTXSID0025654 
3 WS     429174  Arif et al. 2017 1             PAH   Weight DTXSID4020878 
4 WS     429174  Arif et al. 2017 1             PAH   Weight DTXSID4063030 
5 WS     429174  Arif et al. 2017 1             PAH   Weight DTXSID50187178
6 WS     429174  Arif et al. 2017 1             PAH   Weight DTXSID6063143 
  Name                 Value
1 1-Methylphenanthrene NA   
2 1-Methylpyrene       NA   
3 2-Methylnaphthalene  NA   
4 2-Methylpyrene       NA   
5 4-Methylpyrene       NA   
6 5-Methylchrysene     NA

Sample HAWC_ID Study                Sample_Number Class Metric DTXSID       
1 WF     1289737 Campbell et al. 2024 1             PAH   Weight DTXSID1074759
2 WF     1289737 Campbell et al. 2024 1             PAH   Weight DTXSID9020877
3 WF     1289737 Campbell et al. 2024 1             PAH   Weight DTXSID0029187
4 WF     1289737 Campbell et al. 2024 1             PAH   Weight DTXSID4020878
5 WF     1289737 Campbell et al. 2024 1             PAH   Weight DTXSID3061134
6 WF     1289737 Campbell et al. 2024 1             PAH   Weight DTXSID3023845
  Name                    Value
1 1-Methylchrysene        NA   
2 1-Methylnaphthalene     NA   
3 2,6-Dimethylnaphthalene NA   
4 2-Methylnaphthalene     NA   
5 9-Methylanthracene      NA   
6 Acenaphthylene          NA

In [4]:
dim(ws_df)
dim(wf_df)

[1] 4391    9

[1] 3795    9

Starting with 4391 WS records and 3795 WF records. 

In [5]:
# combining dfs
combined_df = rbind(ws_df, wf_df)
head(combined_df)

Sample HAWC_ID Study            Sample_Number Class Metric DTXSID        
1 WS     429174  Arif et al. 2017 1             PAH   Weight DTXSID6025648 
2 WS     429174  Arif et al. 2017 1             PAH   Weight DTXSID0025654 
3 WS     429174  Arif et al. 2017 1             PAH   Weight DTXSID4020878 
4 WS     429174  Arif et al. 2017 1             PAH   Weight DTXSID4063030 
5 WS     429174  Arif et al. 2017 1             PAH   Weight DTXSID50187178
6 WS     429174  Arif et al. 2017 1             PAH   Weight DTXSID6063143 
  Name                 Value
1 1-Methylphenanthrene NA   
2 1-Methylpyrene       NA   
3 2-Methylnaphthalene  NA   
4 2-Methylpyrene       NA   
5 4-Methylpyrene       NA   
6 5-Methylchrysene     NA

In [6]:
length(unique(combined_df$Name))
length(unique(ws_df$Name))
length(unique(wf_df$Name))

[1] 151

[1] 93

[1] 114

114 unique metals/chemicals total.

# Filter #1: Chemical/Metal Background Filters

Determining what variables have missing data for all variables delineated based on the dfs the data will be split into and analyzed. 

A variable (ie. `Name` which represents a chemical or metal) will be retained if >= 20% of its experimental data points were measurable values. In other words, chemicals or metals were excluded if they had NAs (not evaluated) or had non-detect values across >= 80% of the samples.

In [7]:
`%notin%` <- Negate(`%in%`)

sample_type_presence_df = combined_df %>%
    # if an experimental value is there count it as being "present"
    mutate(count = ifelse(Value %notin% c('NA', 'ND'), 1, 0)) %>%
    # determining which have at least 20% experimental values for each chemical/metal
    group_by(Metric, Sample, Name) %>%
    # calculating the percentage of experimental data points
    reframe(Variable_Presence_Percentage = (sum(count)/n()) * 100) %>%
    filter(Variable_Presence_Percentage >= 20) 

head(sample_type_presence_df)

Metric Sample Name                   Variable_Presence_Percentage
1 Volume WF     1,2,4-Trimethylbenzene 100.00000                   
2 Volume WF     Acenaphthene            21.05263                   
3 Volume WF     Acenaphthylene         100.00000                   
4 Volume WF     Anthracene              42.10526                   
5 Volume WF     Arsenic                100.00000                   
6 Volume WF     Benz(a)anthracene      100.00000

In [8]:
# filtering the original df
filter1_df = inner_join(sample_type_presence_df[,1:3], combined_df) 

head(filter1_df)

Joining with `by = join_by(Metric, Sample, Name)`


Metric Sample Name                   HAWC_ID Study                  
1 Volume WF     1,2,4-Trimethylbenzene 1289926 Wang et al. 2024       
2 Volume WF     1,2,4-Trimethylbenzene 1306371 Ketcherside et al. 2024
3 Volume WF     Acenaphthene           1289895 Silberstein et al. 2023
4 Volume WF     Acenaphthene           1286055 Simms et al. 2021      
5 Volume WF     Acenaphthene           1285830 Navarro et al. 2019    
6 Volume WF     Acenaphthene           1285830 Navarro et al. 2019    
  Sample_Number Class DTXSID        Value
1 1             VOC   DTXSID6021402 730  
2 1             VOC   DTXSID6021402 130  
3 1             PAH   DTXSID3021774 0.16 
4 1             PAH   DTXSID3021774 NA   
5 1             PAH   DTXSID3021774 2    
6 2             PAH   DTXSID3021774 2

In [9]:
dim(combined_df)
dim(filter1_df)

[1] 8186    9

[1] 6548    9

Started with 8186 records, 1638 were removed, leaving 6548. 

# Filter #2: Study Background Filter

A study will be retained if >= 20% of its experimental data points were measurable values. In other words, chemicals were excluded if they were NAs (not evaluated) or had non-detect values across < 20% of the samples.

In [10]:
study_type_presence_df = filter1_df %>%
    # if an experimental value is there count it as being "present"
    mutate(count = ifelse(Value %notin% c('NA', 'ND'), 1, 0)) %>%
    # determining which have at least 20% experimental values for each chemical/metal
    group_by(Metric, Class, HAWC_ID) %>%
    # calculating the percentage of experimental data points
    reframe(Study_Presence_Percentage = (sum(count)/n()) * 100) %>%
    filter(Study_Presence_Percentage >= 20) 

head(study_type_presence_df)

Metric Class HAWC_ID Study_Presence_Percentage
1 Volume Metal  266980 100                      
2 Volume Metal  266987 100                      
3 Volume Metal  452289 100                      
4 Volume Metal  885041 100                      
5 Volume Metal  993956 100                      
6 Volume Metal 1070627 100

In [11]:
filter2_df = inner_join(study_type_presence_df[,1:3], filter1_df)

head(filter2_df)

Joining with `by = join_by(Metric, Class, HAWC_ID)`


Metric Class HAWC_ID Sample Name      Study                 Sample_Number
1 Volume Metal 266980  WS     Lead      Stockfelt et al. 2012 1            
2 Volume Metal 266980  WS     Lead      Stockfelt et al. 2012 2            
3 Volume Metal 266980  WS     Potassium Stockfelt et al. 2012 1            
4 Volume Metal 266980  WS     Potassium Stockfelt et al. 2012 2            
5 Volume Metal 266980  WS     Zinc      Stockfelt et al. 2012 1            
6 Volume Metal 266980  WS     Zinc      Stockfelt et al. 2012 2            
  DTXSID Value
1 Pb     170  
2 Pb     400  
3 K      9700 
4 K      8800 
5 Zn     2400 
6 Zn     3100

In [12]:
dim(filter1_df)
dim(filter2_df)

[1] 6548    9

[1] 6426    9

122 more records were removed.

# Filter #3: Chemical/Metal Background Filter

In the intial filter, WS and WF samples were combined. However, this time variables (ie. metal or chemical) will be split based on their `Metric` (ie. volume or weight) and then retained if that variable is found in both woodsmoke and wildfire samples within volume or weight samples.

In [13]:
split_filtered_df = filter2_df %>%
    group_by(Sample, Metric) %>%
    group_split

split_ws_vol_df = split_filtered_df[[3]]
split_ws_weight_df = split_filtered_df[[4]]
split_wf_vol_df = split_filtered_df[[1]]
split_wf_weight_df = split_filtered_df[[2]]

In [14]:
# seeing how many unique variables are in each df and if they're consistent in each file
length(unique(split_ws_vol_df$Name))
length(unique(split_wf_vol_df$Name))
length(unique(split_ws_weight_df$Name))
length(unique(split_wf_weight_df$Name))

[1] 77

[1] 57

[1] 51

[1] 89

In [15]:
# they're not so first getting variables that are in weight or volume samples
consistent_wf_vol_df = split_wf_vol_df %>%
    filter(Name %in% unique(split_ws_vol_df$Name))
consistent_ws_vol_df = split_ws_vol_df %>%
    filter(Name %in% consistent_wf_vol_df$Name)
consistent_wf_weight_df = split_wf_weight_df %>%
    filter(Name %in% unique(split_ws_weight_df$Name))
consistent_ws_weight_df = split_ws_weight_df %>%
    filter(Name %in% consistent_wf_weight_df$Name)

length(unique(consistent_wf_vol_df$Name))
length(unique(consistent_ws_vol_df$Name))
length(unique(consistent_wf_weight_df$Name))
length(unique(consistent_ws_weight_df$Name))

[1] 37

[1] 37

[1] 37

[1] 37

There were 77, 57, 51 and 89 woodsmoke weight, woodsmoke volume, wildfire weight, and wildfire volume samples, respectively. 37 variables were common between volume samples and 37 were common between weight records and will be retained.

In [16]:
# recombining data
vol_df = rbind(consistent_ws_vol_df, consistent_wf_vol_df)
weight_df = rbind(consistent_ws_weight_df, consistent_wf_weight_df)

head(vol_df)

Metric Class HAWC_ID Sample Name      Study                 Sample_Number
1 Volume Metal 266980  WS     Lead      Stockfelt et al. 2012 1            
2 Volume Metal 266980  WS     Lead      Stockfelt et al. 2012 2            
3 Volume Metal 266980  WS     Potassium Stockfelt et al. 2012 1            
4 Volume Metal 266980  WS     Potassium Stockfelt et al. 2012 2            
5 Volume Metal 266980  WS     Zinc      Stockfelt et al. 2012 1            
6 Volume Metal 266980  WS     Zinc      Stockfelt et al. 2012 2            
  DTXSID Value
1 Pb     170  
2 Pb     400  
3 K      9700 
4 K      8800 
5 Zn     2400 
6 Zn     3100

# Square Root Imputation

Taking the square root of the min value to impute MNAR data.

In [17]:
# first determining number of ND and NA values in each df
vol_df %>%
    filter(Value %in% c('NA', 'ND')| grepl("<", Value)) %>%
    group_by(Value) %>%
    summarize(n())

weight_df %>%
    filter(Value %in% c('NA', 'ND')| grepl("<", Value)) %>%
    group_by(Value) %>%
    summarize(n())

Value n()
1 NA    45 
2 ND    27

Value    n()
1 <0.00025   1
2 <0.0005    6
3 <0.0006    2
4 <0.0007    3
5 <0.037     1
6 <0.04      1
7 <0.05      1
8 NA       721
9 ND       211

In [18]:
dim(vol_df)
dim(weight_df)

[1] 1049    9

[1] 3248    9

In [19]:
# creating a df for MAR data
mar_weight_df = weight_df %>%
    filter(Value == 'NA')
mar_vol_df = vol_df %>%
    filter(Value == 'NA')

# creating dfs for studies that contain MNAR data
preimputed_lod_weight_df = anti_join(weight_df, mar_weight_df) 
preimputed_vol_df = anti_join(vol_df, mar_vol_df) 

Joining with `by = join_by(Metric, Class, HAWC_ID, Sample, Name, Study,
Sample_Number, DTXSID, Value)`
Joining with `by = join_by(Metric, Class, HAWC_ID, Sample, Name, Study,
Sample_Number, DTXSID, Value)`


In [20]:
# first imputing non detect values for studies that had a limit of detection (LOD) provided
lod_values_df = preimputed_lod_weight_df %>%
    filter(grepl("<", Value)) %>%
    separate(Value, sep = '<', into = c(NA, 'Value')) %>%
    mutate(Value = sqrt(as.numeric(Value)))

dim(lod_values_df)
head(lod_values_df)

# adding these newly imputed values back into the df
preimputed_weight_df = rbind(anti_join(preimputed_lod_weight_df, lod_values_df[,1:8]), lod_values_df)

[1] 15  9

Metric Class HAWC_ID Sample Name                   Study                 
1 Weight PAH    271571 WS     Acenaphthene           Dilger et al. 2016    
2 Weight PAH    821855 WS     Acenaphthene           Erlandsson et al. 2020
3 Weight PAH    821855 WS     Naphthalene            Erlandsson et al. 2020
4 Weight PAH   1045457 WS     Dibenzo(a,h)anthracene Popadic 2018          
5 Weight PAH   1289739 WF     Benzo(GHI)perylene     Campos et al. 2019    
6 Weight PAH   1289739 WF     Benzo(GHI)perylene     Campos et al. 2019    
  Sample_Number DTXSID        Value     
1 1             DTXSID3021774 0.19235384
2 1             DTXSID3021774 0.22360680
3 1             DTXSID8020913 0.20000000
4 1             DTXSID9020409 0.01581139
5 2             DTXSID5023908 0.02645751
6 6             DTXSID5023908 0.02645751

Joining with `by = join_by(Metric, Class, HAWC_ID, Sample, Name, Study,
Sample_Number, DTXSID)`


In [21]:
preimputed_weight_df$Value = as.numeric(preimputed_weight_df$Value)
preimputed_vol_df$Value = as.numeric(preimputed_vol_df$Value)
mar_weight_df$Value = as.numeric(mar_weight_df$Value)
mar_vol_df$Value = as.numeric(mar_vol_df$Value)

# reordering cols
preimputed_weight_df = preimputed_weight_df[,c(3,1,4,6,7,2,8,5,9)]
preimputed_vol_df = preimputed_vol_df[,c(3,1,4,6,7,2,8,5,9)]

head(preimputed_weight_df)

Warning message:
“NAs introduced by coercion”
Warning message:
“NAs introduced by coercion”
Warning message:
“NAs introduced by coercion”
Warning message:
“NAs introduced by coercion”


HAWC_ID Metric Sample Study                 Sample_Number Class DTXSID
1 267105  Weight WS     Danielsen et al. 2011 1             Metal Cd    
2 267105  Weight WS     Danielsen et al. 2011 2             Metal Cd    
3 267105  Weight WS     Danielsen et al. 2011 1             Metal Cr    
4 267105  Weight WS     Danielsen et al. 2011 2             Metal Cr    
5 267105  Weight WS     Danielsen et al. 2011 1             Metal Cu    
6 267105  Weight WS     Danielsen et al. 2011 2             Metal Cu    
  Name     Value
1 Cadmium  29.4 
2 Cadmium    NA 
3 Chromium 21.4 
4 Chromium 21.3 
5 Copper   65.6 
6 Copper   34.4

In [22]:
# creating a df for MNAR data
mnar_weight_df = preimputed_weight_df %>%
    filter(is.na(Value))
mnar_vol_df = preimputed_vol_df %>%
    filter(is.na(Value))

head(mnar_weight_df)

HAWC_ID Metric Sample Study                 Sample_Number Class DTXSID
1 267105  Weight WS     Danielsen et al. 2011 2             Metal Cd    
2 267105  Weight WS     Danielsen et al. 2011 1             Metal Mn    
3 267105  Weight WS     Danielsen et al. 2011 2             Metal Mn    
4 271638  Weight WS     Jalava et al. 2012    1             Metal Pb    
5 271638  Weight WS     Jalava et al. 2012    2             Metal Pb    
6 271638  Weight WS     Jalava et al. 2012    3             Metal Pb    
  Name      Value
1 Cadmium   NA   
2 Manganese NA   
3 Manganese NA   
4 Lead      NA   
5 Lead      NA   
6 Lead      NA

In [23]:
square_root_imputation = function(mnar_df, preimputed_df){
    # """
    # Imputing MNAR values based on the square root of the minimum concentration for each chemical 
    # within a study or metal.
    
    # :param: MNAR dataframe, preimputed dataframe
    # :output: a dataframe with imputed values

    # """

    id_df = mnar_df[is.na(mnar_df$Value),]
    study_ids = unique(id_df$HAWC_ID)
    
    imputed_df = data.frame()
    for (i in 1:length(id_df$HAWC_ID)){
    
        filtered_df = preimputed_df %>%
            filter(HAWC_ID == id_df$HAWC_ID[i], DTXSID == id_df$DTXSID[i])
    
        na_df = filtered_df %>%
            filter(is.na(Value)) %>%
            mutate(Value = sqrt(min(filtered_df$Value, na.rm = TRUE)))
    
        # combining rows of data
        combined_df = rbind(filtered_df %>%
            filter(!is.na(Value)), na_df)
    
        imputed_df = rbind(imputed_df, combined_df) %>%
                        unique()
    }
    return(imputed_df)    
    
}

# calling fn
imputed_vol_df = square_root_imputation(mnar_vol_df, preimputed_vol_df) %>%
    mutate(Value = ifelse(Value == Inf, NA, Value))
imputed_weight_df = square_root_imputation(mnar_weight_df, preimputed_weight_df) %>%
    mutate(Value = ifelse(Value == Inf, NA, Value))
    
head(imputed_weight_df)

Warning message:
“There was 1 warning in `mutate()`.
ℹ In argument: `Value = sqrt(min(filtered_df$Value, na.rm = TRUE))`.
Caused by warning in `min()`:
! no non-missing arguments to min; returning Inf”
Warning message:
“There was 1 warning in `mutate()`.
ℹ In argument: `Value = sqrt(min(filtered_df$Value, na.rm = TRUE))`.
Caused by warning in `min()`:
! no non-missing arguments to min; returning Inf”
Warning message:
“There was 1 warning in `mutate()`.
ℹ In argument: `Value = sqrt(min(filtered_df$Value, na.rm = TRUE))`.
Caused by warning in `min()`:
! no non-missing arguments to min; returning Inf”
Warning message:
“There was 1 warning in `mutate()`.
ℹ In argument: `Value = sqrt(min(filtered_df$Value, na.rm = TRUE))`.
Caused by warning in `min()`:
! no non-missing arguments to min; returning Inf”
Warning message:
“There was 1 warning in `mutate()`.
ℹ In argument: `Value = sqrt(min(filtered_df$Value, na.rm = TRUE))`.
Caused by warning in `min()`:
! no non-missing arguments to min; retur

HAWC_ID Metric Sample Study                 Sample_Number Class DTXSID
1 267105  Weight WS     Danielsen et al. 2011 1             Metal Cd    
2 267105  Weight WS     Danielsen et al. 2011 2             Metal Cd    
3 267105  Weight WS     Danielsen et al. 2011 1             Metal Mn    
4 267105  Weight WS     Danielsen et al. 2011 2             Metal Mn    
5 271638  Weight WS     Jalava et al. 2012    1             Metal Pb    
6 271638  Weight WS     Jalava et al. 2012    2             Metal Pb    
  Name      Value    
1 Cadmium   29.400000
2 Cadmium    5.422177
3 Manganese        NA
4 Manganese        NA
5 Lead             NA
6 Lead             NA

In [24]:
# there are still a few variables that didn't have any values to do the square root imputation
# their imputed values will be based off of distributions from other studies
other_studies_vol_df = semi_join(preimputed_vol_df %>%
                filter(!is.na(Value)), imputed_vol_df %>%
                               filter(is.na(Value)) %>%
    select(2,3,6,7)) %>%
    group_by(DTXSID) %>%
    summarize(Value = sqrt(min(Value)))

other_studies_weight_df = semi_join(preimputed_weight_df %>%
                filter(!is.na(Value)), imputed_weight_df %>%
                               filter(is.na(Value)) %>%
    select(2,3,6,7)) %>%
    group_by(DTXSID) %>%
    summarize(Value = sqrt(min(Value)))

head(other_studies_vol_df)
dim(other_studies_vol_df)
head(other_studies_weight_df)
dim(other_studies_weight_df)

Joining with `by = join_by(Metric, Sample, Class, DTXSID)`
Joining with `by = join_by(Metric, Sample, Class, DTXSID)`


DTXSID        Value     
1 DTXSID0022432 0.24494897
2 DTXSID0023878 0.06557439
3 DTXSID0023907 0.25690465
4 DTXSID2020139 0.09110434
5 DTXSID3021774 0.40000000
6 DTXSID3024104 0.14142136

[1] 7 2

DTXSID        Value     
1 Al            1.31148770
2 Ba            1.00000000
3 Co            0.10000000
4 Cr            0.31622777
5 DTXSID0023907 0.08660254
6 DTXSID3021774 0.06033241

[1] 14  2

7 (volume) and 14 (weight) samples were imputed based on values from other studies.

In [25]:
# removing NAs from the imputed df 
removed_imputed_vol_df = imputed_vol_df %>%
    filter(!is.na(Value))
removed_imputed_weight_df = imputed_weight_df %>%
    filter(!is.na(Value))

# adding in the newly imputed values back in 
all_imputed_vol_df = left_join(imputed_vol_df %>%
        filter(is.na(Value)) %>%
        select(-Value), other_studies_vol_df)

all_imputed_weight_df = left_join(imputed_weight_df %>%
        filter(is.na(Value)) %>%
        select(-Value), other_studies_weight_df)

all_imputed_vol_df = rbind(removed_imputed_vol_df, all_imputed_vol_df)
all_imputed_weight_df = rbind(removed_imputed_weight_df, all_imputed_weight_df)

head(all_imputed_weight_df)

Joining with `by = join_by(DTXSID)`
Joining with `by = join_by(DTXSID)`


HAWC_ID Metric Sample Study                 Sample_Number Class DTXSID
1 267105  Weight WS     Danielsen et al. 2011 1             Metal Cd    
2 267105  Weight WS     Danielsen et al. 2011 2             Metal Cd    
3 271638  Weight WS     Jalava et al. 2012    4             Metal Si    
4 271638  Weight WS     Jalava et al. 2012    1             Metal Si    
5 271638  Weight WS     Jalava et al. 2012    2             Metal Si    
6 271638  Weight WS     Jalava et al. 2012    3             Metal Si    
  Name    Value       
1 Cadmium    29.400000
2 Cadmium     5.422177
3 Silicon 14700.000000
4 Silicon   121.243557
5 Silicon   121.243557
6 Silicon   121.243557

In [26]:
# creating 1 df
mnar_imputed_vol_df = rbind(#imputed data
                               all_imputed_vol_df,
                               # adding in MAR data
                               mar_vol_df,
                               # adding back in non-imputed data
                               anti_join(preimputed_vol_df, all_imputed_vol_df[,1:8]))
mnar_imputed_weight_df = rbind(# imputed data
                               all_imputed_weight_df,
                               # adding in MAR data
                               mar_weight_df,
                               # adding back in non-imputed data
                               anti_join(preimputed_weight_df, all_imputed_weight_df[,1:8]))

head(mnar_imputed_weight_df)
dim(mnar_imputed_vol_df)
dim(mnar_imputed_weight_df)

Joining with `by = join_by(HAWC_ID, Metric, Sample, Study, Sample_Number,
Class, DTXSID, Name)`
Joining with `by = join_by(HAWC_ID, Metric, Sample, Study, Sample_Number,
Class, DTXSID, Name)`


HAWC_ID Metric Sample Study                 Sample_Number Class DTXSID
1 267105  Weight WS     Danielsen et al. 2011 1             Metal Cd    
2 267105  Weight WS     Danielsen et al. 2011 2             Metal Cd    
3 271638  Weight WS     Jalava et al. 2012    4             Metal Si    
4 271638  Weight WS     Jalava et al. 2012    1             Metal Si    
5 271638  Weight WS     Jalava et al. 2012    2             Metal Si    
6 271638  Weight WS     Jalava et al. 2012    3             Metal Si    
  Name    Value       
1 Cadmium    29.400000
2 Cadmium     5.422177
3 Silicon 14700.000000
4 Silicon   121.243557
5 Silicon   121.243557
6 Silicon   121.243557

[1] 1049    9

[1] 3248    9

# Filter #4: Double Checking MAR Records

Making sure that when the data is split based on `Class` and transformed into a wide format, we have at least 20% of the records with numeric values for each metal or chemical. Converting values to 0s and 1s to count number of experimental and NDs previously imputed. Chemicals/metals that don't have at least 20% of values present will be removed.

In [27]:
wider_vol_df = mnar_imputed_vol_df %>%
    # creating sample ids
    unite("SampleID", c("Sample", "HAWC_ID", "Sample_Number"), remove = FALSE) %>%
    select(-c("Name", "Class")) %>%
    pivot_wider(names_from = "DTXSID", values_from = "Value") 

wider_weight_df = mnar_imputed_weight_df %>%
    # creating sample ids
    unite("SampleID", c("Sample", "HAWC_ID", "Sample_Number"), remove = FALSE) %>%
    select(-c("Name", "Class")) %>%
    pivot_wider(names_from = "DTXSID", values_from = "Value")

head(wider_weight_df)

SampleID    HAWC_ID Metric Sample Study                 Sample_Number
1 WS_267105_1 267105  Weight WS     Danielsen et al. 2011 1            
2 WS_267105_2 267105  Weight WS     Danielsen et al. 2011 2            
3 WS_271638_4 271638  Weight WS     Jalava et al. 2012    4            
4 WS_271638_1 271638  Weight WS     Jalava et al. 2012    1            
5 WS_271638_2 271638  Weight WS     Jalava et al. 2012    2            
6 WS_271638_3 271638  Weight WS     Jalava et al. 2012    3            
  Cd        Si         Ca   Ni   ⋯ Al Ba Co Cu   Zn   K      Mg  DTXSID1025649
1 29.400000         NA   NA 14.7 ⋯ NA NA NA 65.6 1030     NA  NA 10           
2  5.422177         NA   NA 10.3 ⋯ NA NA NA 34.4  750     NA  NA 37           
3 28.000000 14700.0000 1640   NA ⋯ NA NA NA   NA  588  30700 162 NA           
4  6.000000   121.2436 1140   NA ⋯ NA NA NA   NA  250  19700  90 NA           
5 11.000000   121.2436 1780   NA ⋯ NA NA NA   NA 1520 242000 403 NA           
6  6.000000   121.2436  784   NA ⋯ NA NA NA   NA  482  25500  67 NA           
  DTXSID4075459 DTXSID8074819
1    NA         NA           
2    NA         NA           
3  93.4         NA           
4 493.9         NA           
5  48.5         NA           
6 942.0         NA

In [28]:
data_percentage = function(wider_dataset){
    # """
    # Calculating the studies or metals/chemicals with less than or equal to 20% of experimental data.

    # :param: wide dataframe
    # :output: a list that contains a dataframe of metals/chemicals and a dataframe of studies

    # """
    
    # converting values to 0s and 1s
    summed_df = wider_dataset %>%
        mutate(across(7:ncol(wider_dataset), ~ifelse(is.na(.x), 0, 1)))

    # identifying metals/chemicals w/ =< 20% of experimental data
    var_percentage_df = data.frame(Percent = colSums(summed_df[,7:ncol(summed_df)])/nrow(summed_df) * 100) %>%
        filter(Percent <= 20)
    # identifying studies w/ =< 20% of experimental data
    study_percentage_df = data.frame(HAWC_ID = wider_dataset$HAWC_ID,
                                     Percent = rowSums(summed_df[,7:ncol(summed_df)])/ncol(summed_df) * 100) %>%
        filter(Percent <= 20)

    return(list(var_percentage_df, study_percentage_df))
}

# calling fn
percentage_vol_df = data_percentage(wider_vol_df)
percentage_weight_df = data_percentage(wider_weight_df)

head(percentage_weight_df[[1]])

Percent  
Ba            17.058824
DTXSID1025649  5.294118
DTXSID8074819  4.117647

In [29]:
# removing those metals/chemicals & studies from the OG df
preimputed_vol_df = wider_vol_df[,colnames(wider_vol_df) %notin% rownames(percentage_vol_df[[1]])] 
preimputed_weight_df = wider_weight_df[,colnames(wider_weight_df) %notin% rownames(percentage_weight_df[[1]])]

preimputed_vol_df = preimputed_vol_df %>% filter(HAWC_ID %notin% percentage_vol_df[[2]]$HAWC_ID) 
preimputed_weight_df = preimputed_weight_df %>% filter(HAWC_ID %notin% percentage_weight_df[[2]]$HAWC_ID)

head(preimputed_weight_df)

SampleID    HAWC_ID Metric Sample Study                 Sample_Number
1 WS_267105_1 267105  Weight WS     Danielsen et al. 2011 1            
2 WS_267105_2 267105  Weight WS     Danielsen et al. 2011 2            
3 WS_271638_4 271638  Weight WS     Jalava et al. 2012    4            
4 WS_271638_1 271638  Weight WS     Jalava et al. 2012    1            
5 WS_271638_2 271638  Weight WS     Jalava et al. 2012    2            
6 WS_271638_3 271638  Weight WS     Jalava et al. 2012    3            
  Cd        Si         Ca   Ni   ⋯ DTXSID0022432 DTXSID3024289 Mn          Al
1 29.400000         NA   NA 14.7 ⋯     NA         392.0          0.4582576 NA
2  5.422177         NA   NA 10.3 ⋯     NA         245.0          0.4582576 NA
3 28.000000 14700.0000 1640   NA ⋯  227.1         892.7         30.0000000 NA
4  6.000000   121.2436 1140   NA ⋯ 1071.0        3285.0         30.0000000 NA
5 11.000000   121.2436 1780   NA ⋯  132.2         591.4        134.0000000 NA
6  6.000000   121.2436  784   NA ⋯ 2229.0        7373.0         30.0000000 NA
  Co Cu   Zn   K      Mg  DTXSID4075459
1 NA 65.6 1030     NA  NA    NA        
2 NA 34.4  750     NA  NA    NA        
3 NA   NA  588  30700 162  93.4        
4 NA   NA  250  19700  90 493.9        
5 NA   NA 1520 242000 403  48.5        
6 NA   NA  482  25500  67 942.0

In [30]:
preimputed_weight_df[,c(1:5,39)] %>%
    filter(HAWC_ID == 1289807)

SampleID      HAWC_ID Metric Sample Study               Mg
1  WS_1289807_3  1289807 Weight WS     Lamberg et al. 2011 NA
2  WS_1289807_4  1289807 Weight WS     Lamberg et al. 2011 NA
3  WS_1289807_5  1289807 Weight WS     Lamberg et al. 2011 NA
4  WS_1289807_6  1289807 Weight WS     Lamberg et al. 2011 NA
5  WS_1289807_9  1289807 Weight WS     Lamberg et al. 2011 NA
6  WS_1289807_10 1289807 Weight WS     Lamberg et al. 2011 NA
7  WS_1289807_11 1289807 Weight WS     Lamberg et al. 2011 NA
8  WS_1289807_12 1289807 Weight WS     Lamberg et al. 2011 NA
9  WS_1289807_1  1289807 Weight WS     Lamberg et al. 2011 NA
10 WS_1289807_2  1289807 Weight WS     Lamberg et al. 2011 NA
11 WS_1289807_7  1289807 Weight WS     Lamberg et al. 2011 NA
12 WS_1289807_8  1289807 Weight WS     Lamberg et al. 2011 NA

In [31]:
# How many metals/chemicals are now left in the dataset?
ncol(preimputed_vol_df) - 6
ncol(preimputed_weight_df) - 6

[1] 25

[1] 34

In [32]:
# How many studies are left?
nrow(preimputed_vol_df)  
nrow(preimputed_weight_df) 

[1] 29

[1] 122

ADD WORDS. 

# RF Imputation

In [33]:
random_forest_imputation = function(dataset){
    # """
    # Using random forest (RF) to generate MAR data.

    # :param: wide dataframe
    # :output: a longer dataframe 

    # """
    
    preimputed_df = dataset[,7:ncol(dataset)]
    
    # setting seed for reproducibility
    set.seed(38)
    
    imputed_RF_object = missForest(as.matrix(preimputed_df))
    imputed_RF_df = imputed_RF_object$ximp
    imputed_final_df = data.frame(cbind(dataset[,1:6], imputed_RF_df)) %>%
        pivot_longer(cols = 7:ncol(dataset), names_to = "DTXSID", values_to = "Value")
    
    return(imputed_final_df)
}

# calling fn
rf_vol_imputed_df = random_forest_imputation(preimputed_vol_df) 
rf_weight_imputed_df = random_forest_imputation(preimputed_weight_df)

rf_imputed_df = rbind(rf_vol_imputed_df, 
                      rf_weight_imputed_df) 

head(rf_imputed_df)
dim(rf_imputed_df)

SampleID     HAWC_ID Metric Sample Study             Sample_Number
1 WF_1289909_1 1289909 Volume WF     Trieu et al. 2024 1            
2 WF_1289909_1 1289909 Volume WF     Trieu et al. 2024 1            
3 WF_1289909_1 1289909 Volume WF     Trieu et al. 2024 1            
4 WF_1289909_1 1289909 Volume WF     Trieu et al. 2024 1            
5 WF_1289909_1 1289909 Volume WF     Trieu et al. 2024 1            
6 WF_1289909_1 1289909 Volume WF     Trieu et al. 2024 1            
  DTXSID        Value      
1 Cr            77.77900000
2 Cu             5.70000000
3 Pb             5.00000000
4 Ti            33.00000000
5 Zn            22.00000000
6 DTXSID0023878  0.06557439

[1] 4873    8

In [34]:
# adding back in name col
final_df = left_join(rf_imputed_df, 
                      combined_df[,c(5,7:8)] %>%
    unique()) 

final_df = final_df[,c(4,2,5,6,9,3,7,10,8)]

head(final_df)
dim(final_df)

Joining with `by = join_by(DTXSID)`


Sample HAWC_ID Study             Sample_Number Class Metric DTXSID       
1 WF     1289909 Trieu et al. 2024 1             Metal Volume Cr           
2 WF     1289909 Trieu et al. 2024 1             Metal Volume Cu           
3 WF     1289909 Trieu et al. 2024 1             Metal Volume Pb           
4 WF     1289909 Trieu et al. 2024 1             Metal Volume Ti           
5 WF     1289909 Trieu et al. 2024 1             Metal Volume Zn           
6 WF     1289909 Trieu et al. 2024 1             PAH   Volume DTXSID0023878
  Name       Value      
1 Chromium   77.77900000
2 Copper      5.70000000
3 Lead        5.00000000
4 Titanium   33.00000000
5 Zinc       22.00000000
6 Anthracene  0.06557439

[1] 4873    9

In [35]:
# exporting
write.xlsx(final_df, paste0(Output,"/", "Processed_Data_", cur_date, ".xlsx"), 
           rowNames = FALSE)